internal/validate/helpers.go

```diff
@@ -79,3 +90,85 @@ func ReadFile(ctx context.Context, fileName string) (string, error) {
 	log.Debugf("Loaded %s", fileName)
 	return string(fileBytes), nil
 }
+
+// PreProcessPolicy fetches policy sources and returns a policy object with
+// pinned SHA/image digest URL where applicable, along with a policy cache object.
+func PreProcessPolicy(ctx context.Context, policyOptions policy.Options) (policy.Policy, *cache.PolicyCache, error) {
+	var policyCache *cache.PolicyCache
+	pinnedPolicyUrls := map[string][]string{}
+	policyCache, ok := PolicyCacheFromContext(ctx)
+	if !ok {
+		log.Debugf("No cache in context, creating new one")
+		policyCache = cache.NewPolicyCache()
+	} else {
+		log.Debugf("Cache found in context")
+	}
+
+	p, err := NewPolicy(ctx, policyOptions)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sources := p.Spec().Sources
+	for i, sourceGroup := range sources {
+		log.Debugf("Fetching policy source group '%+v'\n", sourceGroup.Name)
+		policySources, err := FetchPolicySources(sourceGroup)
+		if err != nil {
+			log.Debugf("Failed to fetch policy source group '%s'!\n", sourceGroup.Name)
+			return nil, nil, err
+		}
+
+		fs := utils.FS(ctx)
+		dir, err := utils.CreateWorkDir(fs)
+		if err != nil {
+			log.Debug("Failed to create work dir!")
+			return nil, nil, err
+		}
+
+		for _, policySource := range policySources {
+			if strings.HasPrefix(policySource.PolicyUrl(), "data:") {
+				continue
+			}
+
+			destDir, err := policySource.GetPolicy(ctx, dir, false)
+			if err != nil {
+				log.Debugf("Unable to download source from %s!", policySource.PolicyUrl())
+				return nil, nil, err
+			}
+			log.Debugf("Downloaded policy source from %s to %s\n", policySource.PolicyUrl(), destDir)
+
+			url := policySource.PolicyUrl()
+
+			if _, found := policyCache.Get(policySource.PolicyUrl()); !found {
+				log.Debugf("Cache miss for: %s, adding to cache", url)
+				policyCache.Set(url, destDir, nil)
```

Comment:
May want to consider adding the pinned version of the `url` to the cache as well.

Reply:
I'll revisit this once I move the `pinnedUrl function`.